jinahub/indexers/searcher/HnswlibSearcher/__init__.py

```diff
@@ -0,0 +1,101 @@
+    ):
+        """
+        Initialize an HnswlibSearcher
+
+        :param default_top_k: get tok k vectors
+        :param distance: distance can be 'l2', 'ip', or 'cosine'
+        :param dump_path: the path to load ids and vecs
+        :param traverse_path: traverse path on docs, e.g. ['r'], ['c']
+        :param ef_construction: defines a construction time/accuracy trade-off
+        :param ef_query:  sets the query time accuracy/speed trade-off
+        :param max_connection: defines tha maximum number of outgoing connections in the graph
+        :param args:
+        :param kwargs:
+        """
+        super().__init__(*args, **kwargs)
+        self.default_top_k = default_top_k
+        self.distance = distance
+        self.default_traversal_paths = default_traversal_paths or ['r']
+        self.ef_construction = ef_construction
+        self.ef_query = ef_query
+        self.max_connection = max_connection
+        self.logger = get_logger(self)
+        dump_path = dump_path or kwargs.get('runtime_args', {}).get('dump_path', None)
+        if dump_path is not None:
+            self.logger.info('Start building "HnswlibSearcher" from dump data')
+            ids, vecs = import_vectors(dump_path, str(self.metas.pea_id))
+            self._ids = np.array(list(ids))
+            self._vecs = np.array(list(vecs))
+            num_dim = self._vecs.shape[1]
+            self._indexer = hnswlib.Index(space=self.distance, dim=num_dim)
+            self._indexer.init_index(max_elements=len(self._vecs), ef_construction=self.ef_construction, M=self.max_connection)
+
+            self._doc_id_to_offset = {}
+            self._load_index(self._ids, self._vecs)
+        else:
+            self.logger.warning(
+                'No data loaded in "HnswlibSearcher". Use .rolling_update() to re-initialize it...'
+            )
+
+    def _load_index(self, ids, vecs):
+        for idx, v in enumerate(vecs):
+            self._indexer.add_items(v.astype(np.float32), idx)
+            self._doc_id_to_offset[ids[idx]] = idx
+        self._indexer.set_ef(self.ef_query)
+
+    @requests(on='/search')
+    def search(self, docs: DocumentArray, parameters: Dict, **kwargs):
+        if not hasattr(self, '_indexer'):
+            self.logger.warning('Querying against an empty index')
+            return
+        top_k = parameters.get('top_k', self.default_top_k)
+        traversal_paths = parameters.get(
+            'traversal_paths', self.default_traversal_paths
+        )
+
+        for doc in docs.traverse_flat(traversal_paths):
+            indices, dists = self._indexer.knn_query(doc.embedding, k=top_k)
+            for idx, dist in zip(indices[0], dists[0]):
+                match = Document(id=self._ids[idx], embedding=self._vecs[idx])
+                match.scores['distance'] = dist
+                doc.matches.append(match)
+
+    @requests(on='/fill_embedding')
+    def fill_embedding(self, query_da: DocumentArray, **kwargs):
```

Comment:
this has not been tested in a `Flow` right? We cannot change the `signature`

Reply:
yes, I will add a test in Flow, what do you mean by changing the signature?

Reply:
`query_da` is not a valid argument

Reply:
Oh, I see, so we should use `docs` right? I will test it in flow first